tests/agent_disconnect_test.go

```diff
@@ -0,0 +1,244 @@
+			_, err = clientRequest(c, server.URL)
+			if err == nil {
+				t.Errorf("expect request using http persistent connections to fail after dialing on a broken connection")
+			}
+		})
+	}
+}
+
+func TestProxy_Agent_Reconnect(t *testing.T) {
+	testcases := []struct {
+		name                string
+		proxyServerFunction func() (proxy, func(), error)
+		clientFunction      func(string, string) (*http.Client, error)
+	}{
+		{
+			name:                "grpc",
+			proxyServerFunction: runGRPCProxyServer,
+			clientFunction:      createGrpcTunnelClient,
+		},
+		{
+			name:                "http-connect",
+			proxyServerFunction: runHTTPConnProxyServer,
+			clientFunction:      createHTTPConnectClient,
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+
+			server := httptest.NewServer(newEchoServer("hello"))
+			defer server.Close()
+
+			stopCh := make(chan struct{})
+
+			proxy, cleanup, err := tc.proxyServerFunction()
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer cleanup()
+
+			runAgent(proxy.agent, stopCh)
+
+			// Wait for agent to register on proxy server
+			wait.Poll(100*time.Millisecond, 5*time.Second, func() (bool, error) {
+				ready, _ := proxy.server.Readiness.Ready()
+				return ready, nil
+			})
+
+			// run test client
+
+			c, err := tc.clientFunction(proxy.front, server.URL)
+			if err != nil {
+				t.Errorf("error obtaining client: %v", err)
+			}
+
+			_, err = clientRequest(c, server.URL)
+
+			if err != nil {
+				t.Errorf("expected no error on proxy request, got %v", err)
+			}
+			close(stopCh)
+
+			// Wait for the agent to disconnect
+			wait.Poll(100*time.Millisecond, 5*time.Second, func() (bool, error) {
+				ready, _ := proxy.server.Readiness.Ready()
+				return !ready, nil
+			})
+
+			// Reconnect agent
+			stopCh2 := make(chan struct{})
+			runAgent(proxy.agent, stopCh2)
+			defer close(stopCh2)
+
+			// Wait for agent to register on proxy server
+			wait.Poll(100*time.Millisecond, 5*time.Second, func() (bool, error) {
+				ready, _ := proxy.server.Readiness.Ready()
+				return ready, nil
+			})
+
+			// Proxy requests should work again after agent reconnects
+			c2, err := tc.clientFunction(proxy.front, server.URL)
+			if err != nil {
+				t.Errorf("error obtaining client: %v", err)
+			}
+
+			_, err = clientRequest(c2, server.URL)
+
+			if err != nil {
+				t.Errorf("expected no error on proxy request, got %v", err)
+			}
+		})
+	}
+}
+
+func clientRequest(c *http.Client, addr string) ([]byte, error) {
+	r, err := c.Get(addr)
+
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	defer r.Body.Close()
+
+	return data, nil
+}
+
+func createGrpcTunnelClient(proxyAddr, addr string) (*http.Client, error) {
+	tunnel, err := client.CreateSingleUseGrpcTunnel(proxyAddr, grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	c := &http.Client{
+		Transport: &http.Transport{
+			Dial: tunnel.Dial,
+		},
```

Comment:
nit: can you set the Timeout here, and at line 81, checks that the error is not a timeout? Because otherwise, if this bug regresses, the test will just hang there and makes debugging it difficult.

Reply:
Added.